crates/network/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+                                })
+                            }
+                            ClientStatusPacket::Ping(ping) => {
+                                connection.send_packet(
+                                        rjacraft_protocol::packets::server::ServerStatusPacket::Pong(
+                                            rjacraft_protocol::packets::server::Pong {
+                                                payload: ping.payload
+                                            }
+                                        )
+                                    ).unwrap();
+                                connection.flush_packets();
+                            }
+                        },
+                        Err(e) => {
+                            error!("error: {:?}", e)
+                        }
+                    };
+                }
+                HandshakeState::Login => {}
+                HandshakeState::Configuration => {}
+                HandshakeState::Play => {}
+            }
+        }
+    }
+}
+
+#[derive(Resource, Clone)]
+pub struct SharedNetworkState(Arc<SharedNetworkStateInner>);
+
+struct SharedNetworkStateInner {
+    new_connections_send: Sender<RemoteConnection>,
+    new_connections_recv: Receiver<RemoteConnection>,
+}
+
+#[derive(Component)]
+pub struct RemoteConnection {
+    pub remote_addr: SocketAddr,
+    recv: Receiver<PacketFrame>,
+    send: Sender<BytesMut>,
+    encoder: PacketEncoder,
+    recv_task: JoinHandle<()>,
+    send_task: JoinHandle<()>,
+    pub state: HandshakeState,
+}
+
+impl RemoteConnection {
+    pub fn try_recv(&mut self) -> Result<PacketFrame, TryRecvError> {
+        self.recv.try_recv()
+    }
+
+    pub fn write_packet_bytes(&mut self, bytes: &[u8]) -> anyhow::Result<()> {
+        self.encoder.append_frame(bytes)
+    }
+
+    pub fn flush_packets(&mut self) -> Result<(), TrySendError<BytesMut>> {
+        let bytes = self.encoder.take();
+        if !bytes.is_empty() {
+            self.send.try_send(bytes)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn send_packet<T>(&mut self, packet: T) -> anyhow::Result<()>
+    where
+        T: Encoder + Debug,
+    {
+        self.encoder.append_packet(packet)
+    }
+}
+
+impl Drop for RemoteConnection {
+    fn drop(&mut self) {
+        println!("dropping connection {}", self.remote_addr);
+        _ = self.flush_packets();
+        self.recv_task.abort();
+        self.send_task.abort();
+    }
+}
+
+#[derive(Event, Clone, Debug)]
+pub struct PacketReceivedEvent {
+    pub connection: Entity,
+    pub frame: PacketFrame,
+}
+
+#[derive(Event, Clone, Debug)]
+pub struct DisconnectEvent {
+    pub connection: Entity,
+}
+
+#[derive(Event, Clone, Debug)]
+pub struct HandshakeEvent {
+    connection: Entity,
+    protocol_version: i32,
+    server_address: String,
+    server_port: u16,
+    next_state: HandshakeState,
+}
+
+impl HandshakeEvent {
```

Comment:
I think that you could just make all the fields public to have it consistent with other packet `struct`s and to allow things like structural destructuring.

If you want to enforce some invariant on a field, turn the field's value into a newtype wrapping the original type and prohibiting invalid state.